File: solitaire.ex



defmodule Suit do
  @doc "Defines suits"
  use DiscUnion
  defunion 
    Hearts
  | Diamonds
  | Clubs
  | Spades
end

defmodule Rank do
  @doc "Defines ranks"
  use DiscUnion
  defunion 
    King
  | Queen
  | Jack
  | Ten
  | Nine
  | Eight
  | Seven
  | Six
  | Five
  | Four
  | Three
  | Two
  | Ace
end

# defmodule Card do
#   use Discunion, Suit, Rank
#   def new(rank, suit) do
#     %Card{
#       rank: Rank,
#       suit: Suit
#     }
#   end
# end

defimpl String.Chars, for: Rank do
  @doc "implementing to_string for ranks"
  def to_string(rank) do
    King in _ -> "K"
    Queen in _ -> "Q"
    Jack in _ -> "J"
    Ten in _ -> "T"
    Nine in _ -> "9"
    Eight in _ -> "8"
    Seven in _ -> "7"
    Six in _ -> "6"
    Five in _ -> "5"
    Four in _ -> "4"
    Three in _ -> "3"
    Two in _ -> "2"
    Ace in _ -> "A"
  end
end

defimpl String.Chars, for: Suit do
  @doc "implementing to_string for suits"
  def to_string(suit) do
    Spades in _ -> "S"
    Hearts in _ -> "H"
    Clubs in _ -> "C"
    Diamonds in _ -> "D
  end
end

File: solitaire.hs


data Rank = Ace | Two | Three | Four | Five | Six | Seven | Eight | Nine | Ten | Jack | Queen | King
  deriving Eq

instance Show Rank where
  show (Ace) = "A"
  show (Two) = "2"
  show (Three) = "3"
  show (Four) = "4"
  show (Five) = "5"
  show (Six) = "6"
  show (Seven) = "7"
  show (Eight) = "8"
  show (Nine) = "9"
  show (Ten) = "T"
  show (Jack) = "J"
  show (Queen) = "Q"
  show (King) = "K"

instance Show Suit where
  show (Spades) = "S"
  show (Hearts) = "H"
  show (Clubs) = "C"
  show (Diamonds) = "D"

data Suit = Spades | Hearts | Clubs | Diamonds
  deriving Eq

data Card = Card Rank Suit

instance Eq Card where
  (Card r1 s1) == (Card r2 s2) = (r1 == r2) && (s1 == s2)

instance Show Card where
  show (Card r s) = show r ++ show s

type Deck = [Card]

shuffle :: Deck -> IO Deck
shuffle [] = return []
shuffle xs = do 
  randomPosition <- getStdRandom (randomR (0, length xs - 1))
  let (left, (a:right)) = splitAt randomPosition xs 
  map (a:) (shuffle (left ++ right))
				
deck :: Deck
  [Card i j, i <- [Ace, Two, Three, Four, Five, Six, Seven, Eight, Nine, Ten, Jack, Queen, King], j <- [Spades, Hearts, Clubs, Diamonds]]

pick :: Deck -> IO (Maybe Card, Deck)
pick [] = return (Nothing, [])
pick (x:xs) = return (x,xs)